[전우민] 이슈 상세 화면의 오른쪽 사이드 바 컴포넌트 구현

Woomin-Jeon opened the pull request:
Assignees, Labels, Milestone을 하나로 묶어서 컴포넌트 재사용을 무리해서 하려고 했더니 100줄짜리 컴포넌트가 나와버렸네요...
리팩터링으로 그냥 각각으로 나누어서 구현했는데 마지막에 에러가 폭발해버려서 그냥 git reset 했습니다... 엉엉엉
모두들... 오늘... 수고많으셨습니다... 제 똥코드 보시면서.... 조금만 더 수고해주세요...

Woomin-Jeon committed:
[Fix] 잘못된 경로 수정

Woomin-Jeon committed:
[Feat] useParams를 사용하기 위해 URL 변경 및 임시 컴포넌트 삽입

useParams로 URL에서 해당 이슈아이디를 뽑아내기 위해서 /:issueId로
파라미터를 설정함. 아울러 개발 결과를 보기 위해 임시 컴포넌트를 삽입함

Woomin-Jeon committed:
[Package] 옵셔널 체이닝을 사용하기 위해 해당 모듈 설치 및 적용

옵셔널 체이닝 없이 구현하게 되니 너무 지저분해져서 이를 사용하기 위해
해당 모듈을 설치하고 babel.config.js에 추가함

Woomin-Jeon committed:
[Feat] 사이드바 전체를 감싸는 컴포넌트 구현

사이드 바 그자체인 컴포넌트를 구현함

Woomin-Jeon committed:
[Feat] 사이드 바의 각각의 아이템에 대한 컴포넌트 구현

assignees, labels, milestones와 같은 각각의 아이템들을 감싸는 컴포넌트를
구현함

Woomin-Jeon committed:
[Feat] 사이드바의 해당 아이템 명을 출력해주는 컴포넌트 구현

아이템 명 출력을 세부 컴포넌트로 분리

Woomin-Jeon committed:
[Feat] 사이드 바의 해당 아이템 별 피assign 항목을 출력하는 드롭다운

드롭다운으로 해당 아이템에 따라 assign 가능한 항목을 출력해주는 걸
구현함

Woomin-Jeon committed:
[Feat] 자주 사용될 것 같은 컨버터 함수들을 util 함수로 구현

Email을 바탕으로 닉네임을 추출한다거나, ID를 바탕으로 해당 아이템 전체를
받아오는 함수는 자주 사용될 것 같아서 util함수로 구현함

Woomin-Jeon committed:
[Feat] SVG의 문자열을 뽑아내고자 아이콘이라는 파일을 생성

SVG의 d 프로퍼티의 값이 너무 길어 코드의 가독성을 저해하고 있다고
판단하여 따로 분리하였음

Woomin-Jeon committed:
[Feat] SideBarItemDropdown 컴포넌트에 title props 추가

서버로 요청을 보낼 때 해당 assign이 어떤 항목에 대한 assign인지를
알려주기 위해서 title(Assignees, Labels, Mileston 중 하나)를 추가해주어
알 수 있게 함

Woomin-Jeon committed:
[Feat] Assignee와 Label 지정에 대해 서버와 연결 완료

issueAPI에 있던 update 함수를 사용하여 Assignee와 Label에 한해 지정 혹은
이미 지정되어있다면 지정해제가 가능하도록 구현함.

Woomin-Jeon committed:
[Feat] 마일스톤 지정 및 지정해제에 관한 서버 통신 완료

이제 마일스톤을 지정하거나 이미 지정된 것에 대해 해제할 수 있음

Woomin-Jeon committed:
[Fix] 마일스톤이 지정되지 않았을 때 디폴트 메세지 문제 해결

마일스톤이 지정되지 않았을 때 디폴트 메세지가 생기지 않는 문제가
있어서 조건을 하나 더 걸어줌

Woomin-Jeon committed:
[Feat] 서버 통신 이후 인메모리 상태 업데이트

서버 통신이후에 변경사항이 렌더링 되도록 인메모리 상태도 업데이트함

Woomin-Jeon committed:
[Fix] 불필요한 바벨 설정 제거 (옵셔널 체이닝)

없어도 잘 돌아가기 때문에 제거

Woomin-Jeon committed:
[Refactor] if문을 객체의 매핑으로 변경

if문이 너무 지저분해서 리팩터링함

Woomin-Jeon committed:
[Refactor] 데이터가 빈 경우 예외 처리

매번 데이터가 불러왔는지 아닌지 각각에 대해 삼항연산자나
옵셔널체이닝으로 확인했었는데, 한번에 깔끔하게 하는 게 나은 것 같아서
변경함

kkirico commented on client/src/components/issueDetail/SideBar.jsx:
```diff
@@ -0,0 +1,57 @@
+  if (!issues[0] || !labels[0] || !users[0] || !milestones[0]) {
+    return <div />;
+  }
+
```
예외처리 하셨다는게 이 부분이군요! 훨씬 깔끔하고 좋네요

Eunsol0410 replied:
와우!!! 이렇게 깔끔하다면 굳이 하나로 Store를 안 묶어도 될 것 같습니다. 👍

Woomin-Jeon replied:
아휴 참 칭찬 감사합니닿ㅎㅎㅎ

kkirico reviewed (approved):
말씀하신 것과 다르게 너무 이해하기 쉽게 짜셨는데요??
sidebarItemDropdown 파일이 구조가 아쉽긴한데 다른 아이디어가없네요 저도ㅜㅠ
재사용가능한 구조 생각하시면서 코딩하느라 너무 고생 많으셨습니다!

Eunsol0410 commented on client/src/components/issueDetail/SideBarItemDropdown.jsx:
```diff
@@ -0,0 +1,97 @@
+  const { issueId } = useParams();
+  const { issues, dispatch } = useContext(IssuesContext);
+
+  const isAlreadyAssigned = (id) => assigned.find((base) => base.id === id);
```
label이나 asignee는 복수인데, isAlreadyAssigned를 하나만 받아도 문제가 없는지 궁금합니다!

Woomin-Jeon replied:
SideBarItemDropdown 컴포넌트는 드롭다운 아이템 하나하나를 의미하는 거라 괜찮습니다!
그니까 Labels 버튼을 누르면 나오는 목록들 (bug, feature, refactoring, ...등등)의 각 항목(bug라던지)이라고 보시면 될 것같아요!

Eunsol0410 commented on client/src/icons/icons.js:
```diff
@@ -0,0 +1,4 @@
+export const settingIcon = 'M7.429 1.525a6.593 6.593 0 011.142 0c.036.003.108.036.137.146l.289 1.105c.147.56.55.967.997 1.189.174.086.341.183.501.29.417.278.97.423 1.53.27l1.102-.303c.11-.03.175.016.195.046.219.31.41.641.573.989.014.031.022.11-.059.19l-.815.806c-.411.406-.562.957-.53 1.456a4.588 4.588 0 010 .582c-.032.499.119 1.05.53 1.456l.815.806c.08.08.073.159.059.19a6.494 6.494 0 01-.573.99c- ...
+
```
오!! icon 분리하셨군요. 감사합니다.

Eunsol0410 commented on client/src/utils/utils.js:
```diff
@@ -11,3 +11,9 @@ export const deleteStoreItem = (prevState, payload) => {
 };
+
+export const getNicknameByEmail = (email) => email.split('@')[0];
+export const getItemById = (items, id) => {
```
저도 이걸 util로 빼면 좋겠다 생각했는데, 감사합니다~

Eunsol0410 reviewed (approved):
우민님 말씀이 이제 이해되었습니다... ㅎㅎ
컴포넌트 재사용을 하려고 하니... 코드가 길어지는 문제가 있네요
그래도 충분히 이해할 수 있는 것 같습니다!
정말 수고 많으셨습니다 😄

enhakkore commented on client/src/components/issueDetail/SideBarItemDropdown.jsx:
```diff
@@ -0,0 +1,97 @@
+
+      actions[type]();
+      dispatch({ type: 'UPDATE', payload: targetIssue });
+      issueAPI.update({ id: issueId, assignee: { type, id } });
```
API 호출결과가 정상이면 dispatch 하는것이 좋을것 같습니다 😄

Woomin-Jeon replied:
앗 고려를 못했네요! 수정하도록 하겠습니다! :+1:

enhakkore reviewed (approved):
고생하셨습니다 👍

Woomin-Jeon committed:
[Fix] 서버의 응답을 받은 뒤 클라이언트 데이터 변경하도록 수정

서버로부터 제대로 완료했다는 응답을 받고(요청을 성공하지 못하면 null이
반환됨) 그 뒤에 dispatch 로직을 통해 클라이언트 상태를 변경하도록 로직을
수정함

enhakkore committed:
Merge branch 'feature/issue-detail' into feature/issue-detail
